swig/python/codegen/codegen.py

```diff
@@ -0,0 +1,1172 @@
+                        # proceed
+                        self.write(py_func)
+              
+            # message output arg
+            elif t.value.id == self._output_message_var:
+                # check for legit FGPU function calls and translate
+                self.write("FLAMEGPU->message_out.")
+                if t.attr in self.fgpu_output_msg_funcs: 
+                    # proceed
+                    self.write(t.attr)
+                else:
+                    # possible getter setter type function
+                    py_func = self._deviceVariableFunctionName(t, ["setVariable"])
+                    if not py_func:
+                        self.RaiseError(t, f"Function '{t.attr}' does not exist in '{self._output_message_var}' message output object")
+                    # proceed
+                    self.write(py_func)
+                
+                    
+            
+            # math functions (try them in raw function call format) or constants
+            elif t.value.id == "math":
+                self.write(t.attr)
+            # numpy types
+            elif t.value.id == "numpy" or t.value.id == "np":
+                if t.attr in self.numpytypes:
+                    self.write(f"static_cast<{self.numpytypes[t.attr]}>")
+                else: 
+                    self.RaiseError(t, f"Unsupported numpy type {t.attr}")
+            # allow any call on any locals (too many cases to enforce without type checking)
+            elif t.value.id in self._locals:
+                self.write(f"{t.value.id}.{t.attr}")
+            else:
+                self.RaiseError(t, f"Global '{t.value.id}' identifier not supported")
+        # Call is a very nested situation which can occur only on macro environment properties. E.g. 'pyflamegpu.environment.getMacroPropertyInt('a').exchange(10)'
+        elif isinstance(t.value, ast.Call):
+            # handle the call by recursively calling this function to do the depth first execution of pyflamegpu.environment.getMacroPropertyInt('a')
+            self.dispatchMemberFunction(t.value.func, t.value)
+            # check that the handler was actually for macro environment 
+            if t.value.call_type != "MacroEnvironment" :
+                self.RaiseError(t, f"Function call {t.attr} is not supported")
+            # now append the outer call by making sure the thing been called is a valid macro env function
+            if not t.attr in self.fgpu_env_macro_funcs:
+                self.RaiseError(t, f"Function {t.attr} is not a valid macro environment function")
+            # write inner call args
+            self.write("(")
+            self._CallArguments(t.value)
+            self.write(")")
+            # write outer function (call args will be completed by _Call)
+            self.write(f".{t.attr}")
+            
+   
+        else:
+            self.RaiseError(t, "Unsupported function call syntax")
+     
+    
```

Comment:
There are unneccesary blank lines that could be cleaned up, though arguably we don't have python linting rules set up but maybe we should now we have more python (see https://github.com/FLAMEGPU/FLAMEGPU2/issues/666)

Reply:
manually fixed these